File: src/test/kotlin/com/developersam/primitivize/integration/Runtime.kt
package com.developersam.primitivize.integration

import com.developersam.primitivize.runtime.RuntimeFunction
import com.developersam.primitivize.runtime.RuntimeLibrary

/**
 * The simple runtime.
 */
@Suppress(names = ["UNUSED_PARAMETER"])
object Runtime : RuntimeLibrary {

    @RuntimeFunction
    @JvmStatic
    fun memsize(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun defense(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun offense(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun size(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun energy(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun pass(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun tag(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun posture(): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun waitFor(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun forward(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun backward(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun left(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun right(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun eat(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun attack(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun grow(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun bud(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun mate(): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun tag(i: Int): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun serve(i: Int): Unit = Unit

    @RuntimeFunction
    @JvmStatic
    fun nearby(i: Int): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun ahead(i: Int): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun random(i: Int): Int = 0

    @RuntimeFunction
    @JvmStatic
    fun smell(): Int = 0

}

File: rust-rewrite/src/runtime.rs
use crate::ast::{ExpressionStaticType, FunctionType};
use im::{hashmap, HashMap};

pub fn get_critter_world_runtime() -> HashMap<String, FunctionType> {
  hashmap! {
    String::from("memsize") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("defense") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("offense") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("size") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("energy") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("pass") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("posture") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },

    String::from("waitFor") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("forward") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("backward") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("left") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("right") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("eat") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("attack") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("eat") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("attack") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("grow") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("bud") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("mate") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },
    String::from("serve") => FunctionType { argument_types: vec![ExpressionStaticType::IntType], return_type: ExpressionStaticType::VoidType },
    String::from("bud") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::VoidType },

    String::from("nearby") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("ahead") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("random") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
    String::from("smell") => FunctionType { argument_types: vec![], return_type: ExpressionStaticType::IntType },
  }
}
